I'm not entirely sure why the tests are failing 
```pycon
>>> class B:
...     y = 0
...     __slots__ = ('z', 'foo')
... 
>>> B().fod = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'B' object has no attribute 'fod' and no __dict__ for setting new attributes. Did you mean: 'foo'?
```
AFAICT this should be the error that you get? Unless it's failing due to not being printed.

--- a/Lib/test/test_class.py
+++ b/Lib/test/test_class.py
@@ -2,7 +2,6 @@
 
 import unittest
 
-
 testmeths = [
 
 # Binary operations
@@ -641,6 +640,14 @@ class ClassTests(unittest.TestCase):
         class B:
             y = 0
             __slots__ = ('z',)
+        class C:
+            __slots__ = ("y",)
+
+            def __setattr__(self, name, value) -> None:
+                if name == "z":
+                    super().__setattr__("y", 1)
+                else:
+                    super().__setattr__(name, value)
 
         error_msg = "'A' object has no attribute 'x'"
         with self.assertRaisesRegex(AttributeError, error_msg):
@@ -656,6 +663,11 @@ class ClassTests(unittest.TestCase):
         with self.assertRaisesRegex(
             AttributeError,
             "'B' object has no attribute 'x' and no __dict__ for setting new attributes"
+        ):
+            B().x = 0
+        with self.assertRaisesRegex(
+            AttributeError,
+            "'C' object has no attribute 'x'"
         ):
             B().x = 0
 

--- a/Lib/test/test_class.py
+++ b/Lib/test/test_class.py
@@ -640,7 +640,7 @@ class ClassTests(unittest.TestCase):
             pass
         class B:
             y = 0
-            __slots__ = ('z', 'foo')
+            __slots__ = ('z',)
 
         error_msg = "'A' object has no attribute 'x'"
         with self.assertRaisesRegex(AttributeError, error_msg):
@@ -658,14 +658,6 @@ class ClassTests(unittest.TestCase):
             "'B' object has no attribute 'x' and no __dict__ for setting new attributes"
         ):
             B().x = 0
-
-        with self.assertRaisesRegex(
-            AttributeError,
-            "'B' object has no attribute 'fod' and no "
-            r"__dict__ for setting new attributes. Did you mean: 'foo'\?"
-            ""
-        ):
-            B().fod = 1
 
         error_msg = "'B' object attribute 'y' is read-only"
         with self.assertRaisesRegex(AttributeError, error_msg):

--- a/Objects/object.c
+++ b/Objects/object.c
@@ -1545,10 +1545,17 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
         }
         if (dictptr == NULL) {
             if (descr == NULL) {
-                PyErr_Format(PyExc_AttributeError,
-                            "'%.100s' object has no attribute '%U' and no "
-                            "__dict__ for setting new attributes",
-                            tp->tp_name, name);
+                if (tp->tp_setattro == PyObject_GenericSetAttr) {
+                    PyErr_Format(PyExc_AttributeError,
+                                "'%.100s' object has no attribute '%U' and no "
+                                "__dict__ for setting new attributes",
+                                tp->tp_name, name);
+                }
+                else {
+                    PyErr_Format(PyExc_AttributeError,
+                                "'%.100s' object has no attribute '%U'",
+                                tp->tp_name, name);
+                }
                 set_attribute_error_context(obj, name);
             }
             else {

--- a/Objects/object.c
+++ b/Objects/object.c
@@ -1579,8 +1579,7 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
         }
         else {
             PyErr_Format(PyExc_AttributeError,
-                         "'%.100s' object has no attribute '%U' and no "
-                         "__dict__ for setting new attributes",
+                         "'%.100s' object has no attribute '%U'",
                          tp->tp_name, name);
         }
         set_attribute_error_context(obj, name);

